src/compiler/compile/utils/invalidate.ts

```diff
@@ -0,0 +1,65 @@
+import Component from '../Component';
+import MagicString from 'magic-string';
+import { Node } from '../../interfaces';
+import { nodes_match } from '../../utils/nodes_match';
+import { Scope } from './scope';
+
+export function invalidate(component: Component, scope: Scope, code: MagicString, node: Node, names: Set<string>) {
+	const [head, ...tail] = Array.from(names).filter(name => {
+		const owner = scope.find_owner(name);
+		if (owner && owner !== component.instance_scope) return false;
+
+		const variable = component.var_lookup.get(name);
+
+		return variable && (
+			!variable.hoistable &&
+			!variable.global &&
+			!variable.module &&
+			(
+				variable.referenced ||
+				variable.is_reactive_dependency ||
+				variable.export_name
+			)
+		);
+	});
+
+	if (head) {
+		component.has_reactive_assignments = true;
+
+		if (node.operator === '=' && nodes_match(node.left, node.right) && tail.length === 0) {
+			code.overwrite(node.start, node.end, component.invalidate(head));
+		} else {
+			let suffix = ')';
+
+			if (head[0] === '$') {
+				code.prependRight(node.start, `${component.helper('set_store_value')}(${head.slice(1)}, `);
```

Comment:
Is this code being generated in a place where it won't go through the `@sigil` replacement? Is that why you're not using `@set_store_value`?

Reply:
Yeah, it looks like this is the case, but I still don't really have any intuition over which bits of code are going to get run through the sigil replacement.

Reply:
yep. not sure why exactly, I forget how that stuff works. if we follow through on my secret [code-red](https://github.com/Rich-Harris/code-red) plans it won't be an issue in future...